File: android/src/main/java/me/carda/awesome_notifications/awesome_notifications_core/managers/LifeCycleManager.java
package me.carda.awesome_notifications.awesome_notifications_core.managers;

import me.carda.awesome_notifications.awesome_notifications_core.logs.Logger;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.lifecycle.ProcessLifecycleOwner;

import me.carda.awesome_notifications.awesome_notifications_core.AwesomeNotifications;
import me.carda.awesome_notifications.awesome_notifications_core.enumerators.NotificationLifeCycle;
import me.carda.awesome_notifications.awesome_notifications_core.listeners.AwesomeLifeCycleEventListener;

public class LifeCycleManager implements LifecycleObserver {

    private static final String TAG = "LifeCycleManager";

    protected static NotificationLifeCycle appLifeCycle = NotificationLifeCycle.AppKilled;
    public static NotificationLifeCycle getApplicationLifeCycle(){
        return appLifeCycle;
    }

    // ************** SINGLETON PATTERN ***********************

    static LifeCycleManager instance;

    private LifeCycleManager(){
    }

    public static LifeCycleManager getInstance() {
        if (instance == null) {
            instance = new LifeCycleManager();
        }
        return instance;
    }

    // ********************************************************

    // ************** OBSERVER PATTERN ************************

    List<AwesomeLifeCycleEventListener> listeners = new ArrayList<>();

    public LifeCycleManager subscribe(@NonNull AwesomeLifeCycleEventListener listener) {
        listeners.add(listener);
        return this;
    }

    public LifeCycleManager unsubscribe(@NonNull AwesomeLifeCycleEventListener listener) {
        listeners.remove(listener);
        return this;
    }

    public void notify(@NonNull NotificationLifeCycle lifeCycle) {
        for (AwesomeLifeCycleEventListener listener : listeners)
            listener.onNewLifeCycleEvent(lifeCycle);
    }

    // ********************************************************

    boolean hasNotStarted = true;
    public void startListeners(){
        if(hasNotStarted) {
            hasNotStarted = false;
            ProcessLifecycleOwner.get().getLifecycle().addObserver(this);

            if(AwesomeNotifications.debug)
                Logger.d(TAG, "LiceCycleManager listener successfully attached to Android");
        }
    }

    public void stopListeners(){
        if(hasNotStarted) {
            hasNotStarted = false;
            ProcessLifecycleOwner.get().getLifecycle().removeObserver(this);

            if(AwesomeNotifications.debug)
                Logger.d(TAG, "LiceCycleManager listener successfully removed from Android");
        }
    }

    boolean hasGoneForeground = false;
    public void updateAppLifeCycle(NotificationLifeCycle lifeCycle){
        if(appLifeCycle == lifeCycle)
            return;

        hasGoneForeground =
                hasGoneForeground ||
                appLifeCycle == NotificationLifeCycle.Foreground;

        appLifeCycle = lifeCycle;
        notify(appLifeCycle);

        if(AwesomeNotifications.debug)
            Logger.d(TAG, "App is now "+lifeCycle);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    public void onCreated() {
        updateAppLifeCycle(
                hasGoneForeground ?
                        NotificationLifeCycle.Background :
                        NotificationLifeCycle.AppKilled);
    }

    boolean wasNotCreated = true;
    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void onStarted() {
        updateAppLifeCycle(
                hasGoneForeground ?
                        NotificationLifeCycle.Background :
                        NotificationLifeCycle.AppKilled);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    public void onResumed() {
        updateAppLifeCycle(
                NotificationLifeCycle.Foreground);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    public void onPaused() {
        updateAppLifeCycle(
                NotificationLifeCycle.Foreground);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    public void onStopped() {
        updateAppLifeCycle(
                NotificationLifeCycle.Background);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void onDestroyed() {
        updateAppLifeCycle(
                NotificationLifeCycle.AppKilled);
    }

}

File: ios/core_old/managers/LifeCycleManager.swift
//
//  LifeCycleManager.swift
//  Pods
//
//  Created by Rafael Setragni on 05/11/20.
//

import Foundation

public class LifeCycleManager:
    NSObject,
    UIApplicationDelegate {
    
    private let TAG = "LifeCycleManager"
    
    static let _userDefaults = UserDefaults(suiteName: Definitions.USER_DEFAULT_TAG)
    private let referenceKey: String = "currentlifeCycle"
    
    // ************** SINGLETON PATTERN ***********************
    
    static var instance:LifeCycleManager?
    public static var shared:LifeCycleManager {
        get {
            LifeCycleManager.instance =
                LifeCycleManager.instance ?? LifeCycleManager()
            return LifeCycleManager.instance!
        }
    }
    private override init(){
        super.init()
    }
    
    // ************** IOS EVENTS LISTENERS ************************
    
    var _listening = false
    public func startListeners(){
        if _listening {
            return
        }

        _listening = true
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.applicationDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification, object: nil)
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.applicationWillResignActive),
            name: UIApplication.willResignActiveNotification, object: nil)
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.applicationDidEnterBackground),
            name: UIApplication.didEnterBackgroundNotification, object: nil)
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.applicationWillTerminate),
            name: UIApplication.willTerminateNotification, object: nil)
        
        if !SwiftUtils.isRunningOnExtension() {
            currentLifeCycle = .AppKilled
        }
    }
    
    deinit {
        if _listening {
            NotificationCenter.default.removeObserver(self)
        }
    }
    
    // ************** OBSERVER PATTERN ************************
    
    private lazy var listeners = [AwesomeLifeCycleEventListener]()
    
    public func subscribe(listener:AwesomeLifeCycleEventListener) -> Self {
        listeners.append(listener)
        if(AwesomeNotifications.debug){
            Logger.d(TAG, "LiceCycleManager listener successfully attached to iOS")
        }
        return self
    }
    
    public func unsubscribe(listener:AwesomeLifeCycleEventListener) {
        if let index = listeners.firstIndex(where: {$0 === listener}) {
            listeners.remove(at: index)
            if(AwesomeNotifications.debug){
                Logger.d(TAG, "LiceCycleManager listener successfully removed from iOS")
            }
        }
    }
    
    private func notify(lifeCycle: NotificationLifeCycle){
        for listener in listeners {
            listener.onNewLifeCycleEvent(lifeCycle: lifeCycle)
        }
    }
    
    // ********************************************************
    
    private var _oldLifeCycle:NotificationLifeCycle?
    private var _currentLifeCycle:NotificationLifeCycle?
    
    public var currentLifeCycle: NotificationLifeCycle {
        get {
            if SwiftUtils.isRunningOnExtension() {
                if let rawName =
                        LifeCycleManager
                            ._userDefaults?
                            .string(forKey: referenceKey)
                {
                    _currentLifeCycle = EnumUtils.fromString(rawName)
                }
            }
            return _currentLifeCycle ?? NotificationLifeCycle.AppKilled
        }
        set {
            _currentLifeCycle = newValue
            
            LifeCycleManager
                ._userDefaults?
                .setValue(newValue.rawValue, forKey: referenceKey)
            
            if _currentLifeCycle == .Foreground {
                _hasGoneForeground = true
            }
            
            if _currentLifeCycle != _oldLifeCycle {
                _oldLifeCycle = _currentLifeCycle
                notify(lifeCycle: newValue)
                if AwesomeNotifications.debug {
                    Logger.d(TAG, "App is now "+newValue.rawValue)
                }
            }
        }
    }
    
    var _isOutOfFocus = false
    public var isOutOfFocus:Bool {
        get {
            return _isOutOfFocus
        }
    }
    
    var _hasGoneForeground = false
    public var hasGoneForeground:Bool {
        get {
            return _hasGoneForeground
        }
    }
    
    // ******************************  IOS LIFECYCLE EVENTS  ***********************************
    
    public func applicationDidBecomeActive(_ application: UIApplication) {
        currentLifeCycle = .Foreground
        _isOutOfFocus = false
    }
    
    public func applicationWillResignActive(_ application: UIApplication) {
        _isOutOfFocus = true
    }
    
    public func applicationDidEnterBackground(_ application: UIApplication) {
        currentLifeCycle = hasGoneForeground ? .Background : .AppKilled
    }
    
    public func applicationWillTerminate(_ application: UIApplication) {
        currentLifeCycle = .AppKilled
    }
    
}
